ESP8266: Avoid duplicate data sends (5.15)

ccli8 opened the pull request:
### Summary of changes <!-- Required -->

Backport of #12157 to mbed-os-5.15.

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

@michalpasztamobica 

----------------------------------------------------------------------------------------------------------------

michalpasztamobica committed:
ESP8266: add a retry mechanism to avoid duplicate data sends

We are now checking if ESP8266 has confirmed receiving data over serial
port with an undocumented (but existing) "Recv x bytes" message. Next we
are explicitly waiting for an official "SEND OK".

michalpasztamobica committed:
ESP8266: Improve error handling and partial sends

michalpasztamobica committed:
ESP8266: Introduce single send_status instead of two flags

michalpasztamobica committed:
Increase UDP test timeout to 1200s

This is because the ESP8266 is now waiting for SEND OK and takes much
more to complete the send_repeat, and echo_burst tests in RAAS.
(but has higher pass ratio).

ccli8 committed:
ESP8266: Simplify flow control and enable per-socket reconnection

1.  Fix 'spurious close' by adding close() in open(). 'spurious close' gets frequent and cannot ignore when send() changes to asynchronous. User can retry open() until 'spurious close' gets true.
2.  Allow only one actively sending socket because:
    (1) ESP8266 AT packets 'SEND OK'/'SEND FAIL' are not associated with socket ID. No way to tell them.
    (2) In original implementation, ESP8266::send() is synchronous, which implies only one actively sending socket.
3.  Register 'SEND OK'/'SEND FAIL' oobs, like others in ESP8266::ESP8266 constructor. Don't get involved in oob management with send status because ESP8266 modem possibly doesn't reply these packets on error case.
4.  Now that ESP8266::send() changes to asynchronous, drop the code with _parser.recv("SEND OK")/_parser.recv("SEND FAIL"). _parser.recv("SEND OK")/_parser.recv("SEND FAIL") and 'SEND OK'/'SEND FAIL' oobs both consume 'SEND OK'/'SEND FAIL' packets and complicate flow control.

michalpasztamobica committed:
ESP8266: accept partial writes for TCP and clean up code

michalpasztamobica reviewed (approved):
If the fix is required on the 5.15 branch, then I don't mind. Technically this looks sane.

ccli8 committed:
ESP8266: Add retries to avoid spurious close in blocking mode

With send(...) changing to non-block, modem can be busy in sending previous data and 'busy' the current 'AT+CIPCLOSE' command.
In blocking mode, add retries to avoid spurious close to some degree.
This is required to pass GT netsocket-tcp/netsocket-tls tests which expect close(...) to be OK.
